tensorflow/compiler/mlir/tosa/transforms/legalize_common.cc

```diff
@@ -2966,84 +2967,125 @@ llvm::Optional<Value> convertResizeOp(PatternRewriter& rewriter, Operation* op,
   size_t output_height = output_shape[1];
   size_t output_width = output_shape[2];
 
-  double fp_stride_y =
-      static_cast<double>(input_height) / static_cast<double>(output_height);
-  double fp_stride_x =
-      static_cast<double>(input_width) / static_cast<double>(output_width);
-  if (align_corners && output_height > 1) {
-    fp_stride_y = static_cast<double>(input_height - 1) /
-                  static_cast<double>(output_height - 1);
-  }
-  if (align_corners && output_width > 1) {
-    fp_stride_x = static_cast<double>(input_width - 1) /
-                  static_cast<double>(output_width - 1);
-  }
-
-  double fp_offset_y, fp_offset_x;
-  if (half_pixel_centers) {
-    fp_offset_y = fp_stride_y * 0.5f - 0.5f;
-    fp_offset_x = fp_stride_x * 0.5f - 0.5f;
-  } else {
-    fp_offset_y = 0.0f;
-    fp_offset_x = 0.0f;
-  }
+  auto isPowerOfTwoUpscale = [](size_t a, size_t b) {
+    auto isPowerOfTwo = [](size_t x) { return x && (!(x & (x - 1))); };
+    return (a % b == 0) && isPowerOfTwo(a / b);
+  };
+
+  // a. For power of two upscale [OH, OW] = (1 << k) * [IH, IW],
```

Comment:
And these are to get the most accurate results?

Reply:
It's a simplification to allow exact calculation of the output size.
By defining the scaling factor as a ratio of integers so that exact output dimensions can be derived from input dimensions without rounding.

Reply:
Could you add that just above this? (the why we are doing this before the how)

Reply:
Done.